### PR TITLE
Handlers

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change ###

Setup work for handlers on the main branch

### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Core/RectangleTypeConverter.cs [hunk 1]
```diff
@@ -12,7 +12,7 @@ public override object ConvertFromInvariantString(string value)
 			if (value != null)
 			{
 				string[] xywh = value.Split(',');
-				if (   xywh.Length == 4
+				if (xywh.Length == 4
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/RectangleTypeConverter.cs:15]: the reason for this extra spaces is to align the members of the logical AND, like this
```
if (   expr0
    && expr1)
```

which is more readable (for complex or long expressions) than
```
if (expr0
    && expr1)
```

or
```
if (expr0 &&
    expr1)
```

File: Xamarin.Forms.Core/Button.Impl.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using Xamarin.Platform;
+
+namespace Xamarin.Forms
+{
+	public partial class Button : IButton
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Button.Impl.cs:8]: looks like a very short ident. could be GH messing up with the visual...

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.Platform
+{
+	public class PropertyMapper {
+		internal Dictionary<string, (Action<IViewRenderer, IFrameworkElement> Action, bool RunOnUpdateAll)> genericMap = new Dictionary<string, (Action<IViewRenderer, IFrameworkElement> action, bool runOnUpdateAll)> ();
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:9]: we know it's generic, it needs a better name

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			if (virtualView == null)
+				return;
+			foreach (var key in Keys) {
+				UpdateProperty (key, viewRenderer, virtualView);
+			}
+		}
+		public virtual ICollection<string> Keys => genericMap.Keys;
+	}
+
+	public class PropertyMapper<TVirtualView> : PropertyMapper, IEnumerable
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:34]: is there a use case for this to be subclassed (with part of the API being internal) ?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.Platform
+{
+	public class PropertyMapper {
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:8]: is there a use case for this to be subclassed by users (with part of the API being internal) ?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:4]: the mount of Linq, and collection handling, will need to be evaluated for performance on devices

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+		public void Add(string key, Action<IViewRenderer, TVirtualView> action, bool ignoreOnStartup)
+			=>genericMap[key] = ((r, v) => action?.Invoke(r, (TVirtualView)v), ignoreOnStartup);
+
+		
+		
+
+		IEnumerator IEnumerable.GetEnumerator () => genericMap.GetEnumerator ();
+
+		public class ActionMapper<TView>
```
Comment (StephaneDelcroix) [inline src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:91]: if it's public, avoid a nested type

File: Xamarin.Forms.Platform.Android/Xamarin.Forms.Platform.Android.csproj [hunk 8]
```diff
@@ -1,61 +1,17 @@
...
-  <ItemGroup>
-    <ProjectReference Include="..\Xamarin.Forms.Core\Xamarin.Forms.Core.csproj">
-    </ProjectReference>
-    <ProjectReference Include="..\Xamarin.Forms.Platform.Android.FormsViewGroup\Xamarin.Forms.Platform.Android.FormsViewGroup.csproj">
-    </ProjectReference>
-  </ItemGroup>
-  <ItemGroup>
-    <Folder Include="Shapes\" />
+  <ItemGroup >
+    <None Include="**\*.cs" />
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply Xamarin.Forms.Platform.Android/Xamarin.Forms.Platform.Android.csproj:15]: Right!

We're going to gut this project and move the code into extension method/handlers 
As we move the code we'll comment it out so we keep track of all the code we've moved into extension methods/handlers

When we merge up we'll get merge conflicts where code has been commented out and that code will need to be moved to the handler

File: src/Platform.Handlers/samples/Sample/Sample.csproj [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		<None Include="**\*.MaciOS.cs" />
+	</ItemGroup>
+	<ItemGroup Condition="$(TargetFramework.StartsWith('MonoAndroid')) != true ">
+		<Compile Remove="**\*.Android.cs" />
+		<None Include="**\*.Android.cs" />
+	</ItemGroup>
+	<ItemGroup Condition="$(TargetFramework.StartsWith('netstandard')) != true ">
+		<Compile Remove="**\*.Standard.cs" />
+		<None Include="**\*.Standard.cs" />
+	</ItemGroup>
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/samples/Sample/Sample.csproj:31]: 👍  I moved it to the .nuspec folder

If there isn't an alternative multi targeting strategy by the time we're publishing nugets then maybe we can just include this targets file for the time being to help users get started

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Forms/Color.cs [hunk 10]
```diff
@@ -2,12 +2,12 @@
...
 using System.Diagnostics;
 using System.Globalization;
-using Xamarin.Forms.Internals;
+using Xamarin.Platform.Extensions;
 
 namespace Xamarin.Forms
 {
 	[DebuggerDisplay("R={R}, G={G}, B={B}, A={A}, Hue={Hue}, Saturation={Saturation}, Luminosity={Luminosity}")]
-	[TypeConverter(typeof(ColorTypeConverter))]
+	//[TypeConverter(typeof(ColorTypeConverter))]
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Forms/Color.cs:10]: https://github.com/dotnet/maui/issues/228

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Forms/Rect.cs [hunk 11]
```diff
@@ -5,7 +5,7 @@
 namespace Xamarin.Forms
 {
 	[DebuggerDisplay("X={X}, Y={Y}, Width={Width}, Height={Height}")]
-	[TypeConverter(typeof(RectTypeConverter))]
+	//[TypeConverter(typeof(RectTypeConverter))]
 	public struct Rect
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Forms/Rect.cs:9]: Maybe this is the eventual plan? Current plan was to minimize anything breaking and just use our own types and then at some point later on we'll switch over to the System types.

Is this a change that's important to do now?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+namespace Xamarin.Platform
+{
+	public class PropertyMapper {
+		internal Dictionary<string, (Action<IViewRenderer, IFrameworkElement> Action, bool RunOnUpdateAll)> genericMap = new Dictionary<string, (Action<IViewRenderer, IFrameworkElement> action, bool runOnUpdateAll)> ();
+		protected virtual void UpdateProperty (string key, IViewRenderer viewRenderer, IFrameworkElement virtualView)
+		{
+			if (genericMap.TryGetValue(key, out var action))
+			{
+				action.Action?.Invoke(viewRenderer, virtualView);
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:14]: Hmmm possibly? 

> can't we move it to when the action is registered ?

Move what specifically? Just register an empty action? or remove the key if the user sets the action to null?

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/Extensions/RectExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+using System.Linq;
+using System.Text;
+using Xamarin.Forms;
+
+namespace Xamarin.Platform.Core
+{
+	public static class RectExtensions
+	{
+		public static Point Center(this Rectangle rectangle) => new Point(rectangle.X + (rectangle.Width / 2), rectangle.Y + (rectangle.Height / 2));
+		public static void Center(this ref Rectangle rectangle, Point center)
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/Extensions/RectExtensions.cs:12]: It looks like these are used by the CollectionView snap helpers so I'll just remove them for now

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs (hunk 5 above)
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs]: We're subclassing it for the Generic mapper here

https://github.com/xamarin/Xamarin.Forms/blob/83a7ad4f3e5bbdc4a4fbc97965617710e6c9052d/src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs#L34

So yes

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs:16]: We decided to keep this how it is and let "Null" represent setting the action to an noop method

So the breakdowns of how you can interact with this API are


```C#
		static MyCustomButtonHandler()
		{

			// custom behavior
			ButtonHandler.ButtonMapper[nameof(IButton.Text)] = CustomBehavior;

			// resets to default
			ButtonHandler.ButtonMapper[nameof(IButton.Text)] = ButtonHandler.MapText;

			// basically the same as using an empty method
			ButtonHandler.ButtonMapper[nameof(IButton.Text)] = null;
			ButtonHandler.ButtonMapper[nameof(IButton.Text)] = (_, __) => { };
		}
```

File: src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs [hunk 14]
```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Xamarin.Platform
+{
+	public class PropertyMapper {
+		internal Dictionary<string, (Action<IViewRenderer, IFrameworkElement> Action, bool RunOnUpdateAll)> genericMap = new Dictionary<string, (Action<IViewRenderer, IFrameworkElement> action, bool runOnUpdateAll)> ();
+		protected virtual void UpdateProperty (string key, IViewRenderer viewRenderer, IFrameworkElement virtualView)
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Platform.Handlers/src/Xamarin.Platform.Handlers/PropertyMapper.cs]: For now we'll keep this as a string.. I might not fully be understanding your statement here

UpdateProperty isn't going to exist at the BP layer at all. Maybe I'm missing something?

**Comment by rookiejava** [comment]
@PureWeen As for Tizen work, I'll do when this PR is merged. Is it OK?

**Pull request merged**